File: tugas array kuliah/array_fixed.py
import sys

# Konfigurasi Global
MAX = 10
count = 0
# Inisialisasi array 2D fix (10 baris, 2 kolom) dengan None
arr = [[None, None] for _ in range(MAX)]

def main():
    print("be wise when using sawit program")
    print("choose 1 - 8 for operation and 9 for close program")
    print("1. insert at beggining")
    print("2. insert given position")
    print("3. insert at end")
    print("4. delete from beggining")
    print("5. delete given position")
    print("6. delete frome end")
    print("7. delete first occurence")
    print("8. show data")
    print("9. exit")

    choice = 0
    while choice != 9:
        try:
            raw_input = input("9 to exit, your choice : ")
            choice = int(raw_input)
            
            if choice == 1:
                insert_at_beginning()
            elif choice == 2:
                insert_given_position()
            elif choice == 3:
                insert_at_end()
            elif choice == 4:
                delete_from_beginning()
            elif choice == 5:
                delete_given_position()
            elif choice == 6:
                delete_from_end()
            elif choice == 7:
                delete_first_occurence()
            elif choice == 8:
                show_data()
            elif choice == 9:
                break
            else:
                print("input only accept 1-9 ")
        except ValueError:
            print("can't input string! only accept number")

    print("thank you for using sawit program")

def insert_at_beginning():
    global count
    # Geser data dari index 8 ke 9, 7 ke 8, dst.
    for i in range(MAX - 1, 0, -1):
        arr[i][0] = arr[i - 1][0]
        arr[i][1] = arr[i - 1][1]

    arr[0][0] = input("input name: ")
    arr[0][1] = input("input nim: ")

    if count < MAX:
        count += 1

def insert_given_position():
    global count
    if count >= MAX or count < 0:
        print("=== Array is full ===")
        return

    try:
        position = int(input("choose the position (1-10): "))
        
        if 1 <= position <= MAX:
            index = position - 1

            if arr[index][0] is not None:
                print("array sudah ada yang mengisi, Y/N untuk Overwrite: ", end="")
                force = input().strip().upper()
                
                if force == 'Y':
                    arr[index][0] = input("input name: ")
                    arr[index][1] = input("input nim: ")
                elif force == 'N':
                    print("array not overwrite")
                else:
                    print("only accept Y/N")
            else:
                arr[index][0] = input("input name: ")
                arr[index][1] = input("input nim: ")

            if count < MAX:
                count += 1
        else:
            print("range out of the bound")
    except ValueError:
        print("position must be integer")

def insert_at_end():
    global count
    max_index = 9
    last_occupied = -1

    # Cek dari belakang (9 ke 0)
    for i in range(max_index, -1, -1):
        if arr[i][0] is not None:
            last_occupied = i
            break
    
    # Jika index terakhir (9) penuh
    if last_occupied == max_index:
        print("wanna force input (Y/N): ", end="")
        forced = input().strip().upper()

        if forced == 'Y':
            arr[max_index][0] = input("input name: ")
            arr[max_index][1] = input("input nim: ")
        else:
            print("array not overwrite")
    else:
        # Isi slot kosong setelah index terakhir yang terisi
        pos = last_occupied + 1
        arr[pos][0] = input("input name: ")
        arr[pos][1] = input("input nim: ")
        
        if count < MAX:
            count += 1

def delete_from_beginning():
    global count
    if count == 0:
        print("Array is empty")
        return

    # Geser ke kiri (shifting logic)
    for i in range(0, count - 1):
        arr[i][0] = arr[i + 1][0]
        arr[i][1] = arr[i + 1][1]

    # Kosongkan data terakhir yang tersisa setelah geser
    arr[count - 1][0] = None
    arr[count - 1][1] = None
    count -= 1

def delete_given_position():
    global count
    if count == 0:
        print("Array is empty")
        return

    try:
        position = int(input("input position: "))
        
        if 1 <= position <= MAX:
            index = position - 1
            
            if arr[index][0] is None:
                print("Position is already empty")
                return
            
            # Set null tanpa shifting (sesuai kode Java terakhir)
            arr[index][0] = None
            arr[index][1] = None
            count -= 1
        else:
            print("range out of the bound")
    except ValueError:
        print("only accept int")

def delete_from_end():
    global count
    max_index = 9

    # Cek dari belakang untuk hapus yang paling akhir
    for i in range(max_index, -1, -1):
        if arr[i][0] is not None:
            arr[i][0] = None
            arr[i][1] = None
            
            if count > 0:
                count -= 1
            
            print(f"Deleted data at index {i}")
            return

    print("Array is empty")

def delete_first_occurence():
    global count
    if count == 0:
        print("Array is empty")
        return

    nim_input = input("input nim will be delete: ")
    # Hapus semua spasi dari input
    target_nim = nim_input.replace(" ", "")

    index = -1

    # Scan dari depan (0 ke 9)
    for i in range(MAX):
        if arr[i][1] is not None:
            # Hapus spasi dari data array untuk perbandingan
            current_nim = arr[i][1].replace(" ", "")
            if current_nim.lower() == target_nim.lower():
                index = i
                break
    
    if index != -1:
        # Hapus data (set None) tanpa shifting
        arr[index][0] = None
        arr[index][1] = None

        if count > 0:
            count -= 1
        print("Data deleted successfully")
    else:
        print("nim not found")

def show_data():
    print(f"Length array / total data: {count}")
    for i in range(MAX):
        if arr[i][0] is not None:
            print(f"{i + 1} : {arr[i][0]} || {arr[i][1]}")
        else:
            print(f"{i + 1} : [Empty]")

if __name__ == "__main__":
    main()

File: array_fixed.java
import java.util.Scanner;

public class array_fixed {

    static int count = 0;
    static final int MAX = 10;

    public static void main(String[] args){
        Init();
    }

    static void Init(){
        Scanner input = new Scanner(System.in);
        String[][] arr = new String[MAX][2];

        System.out.println("be wise when using sawit program");
        System.out.println("choose 1 - 8 for operation and 9 for close program");
        System.out.println("1. insert at beggining");
        System.out.println("2. insert given position");
        System.out.println("3. insert at end");
        System.out.println("4. delete from beggining");
        System.out.println("5. delete given position");
        System.out.println("6. delete frome end");
        System.out.println("7. delete first occurence");
        System.out.println("8. show data");
        System.out.println("9. exit");

        int choice = 0;
        do{
            System.out.print("9 to exit, your choice : ");
            
            if(input.hasNextInt())
            {
                choice = input.nextInt();
                input.nextLine();
                
                switch(choice)
                {
                    case 1: 
                        InsertAtBeggining(arr, input);
                        ShowData(arr);
                        break;
                    case 2:
                        InsertGivenPosition(arr, input);
                        ShowData(arr);
                        break;
                    case 3:
                        InsertAtEnd(arr, input);
                        ShowData(arr);
                        break;
                    case 4:
                        DeleteFromBeggining(arr);
                        ShowData(arr);
                        break;
                    case 5:
                        DeleteGivenPosition(arr, input);
                        ShowData(arr);
                        break;
                    case 6: 
                        DeleteFromEnd(arr);
                        ShowData(arr);
                        break;
                    case 7:
                        DeleteFirstOccurence(arr, input);
                        ShowData(arr);
                        break;
                    case 8:
                        ShowData(arr);
                        break;
                    case 9:
                        break;
                    default:
                        System.out.println("input only accept 1-9 ");
                }
            } else {
                System.out.println("can't input string! only accept number");
                input.next();
            }
        } while(choice != 9);

        System.out.println("thank you for using sawit program");
        input.close();
    }

    static void InsertAtBeggining(String[][] arr, Scanner input)
    {
        for (int i = MAX - 1; i > 0; i--) {
            arr[i][0] = arr[i - 1][0];
            arr[i][1] = arr[i - 1][1];
        }

        System.out.print("input name: ");
        arr[0][0] = input.nextLine();
        System.out.print("input nim: ");
        arr[0][1] = input.nextLine();

        if (count < MAX) {
            count++;
        }
    }

    static void InsertGivenPosition(String[][] arr, Scanner input){
        if (count >= MAX) {
            System.out.println("=== Array is full ===");
            //return;
        }

        System.out.print("choose the position (1-10): ");
        if (input.hasNextInt()){
            int position = input.nextInt();
            input.nextLine();

            if(position >= 1 && position <= MAX){
                int index = position - 1;

            if(arr[index][0] != null){
                System.out.print("array sudah ada yang mengisi, Y/N untuk Overwrite: ");
                char force = input.next().charAt(0);
                
                switch(Character.toUpperCase(force)){
                    case 'Y':
                        input.nextLine();
                        System.out.print("input name: ");
                        arr[index][0] = input.nextLine();
                        System.out.print("input nim: ");
                        arr[index][1] = input.nextLine();
                        break;
                    case 'N':
                        input.nextLine();
                        System.out.println("array not overwrite");
                        break;
                    default:
                        System.out.println("only accept Y/N");
                }
            } else {
                System.out.print("input name: ");
                arr[index][0] = input.nextLine();
                System.out.print("input nim: ");
                arr[index][1] = input.nextLine();

            }
            if(count < MAX){
                count++;
            }
                
            } else {
                System.out.println("range out of the bound");
            }
        } else {
            System.out.println("position must be integer");
        }
    }

    static void InsertAtEnd(String[][] arr, Scanner input){
        if (count >= MAX) {
            System.out.println("Array is full");
            return;
        }

        System.out.print("input name: ");
        String name = input.nextLine();
        System.out.print("input nim: ");
        String nim = input.nextLine();

        arr[count][0] = name;
        arr[count][1] = nim;

        count++;
    }

    static void DeleteFromBeggining(String[][] arr){
        if (count == 0) {
            System.out.println("Array is empty");
            return;
        }

        for (int i = 0; i < count - 1; i++) {
            arr[i][0] = arr[i + 1][0];
            arr[i][1] = arr[i + 1][1];
        }

        arr[count - 1][0] = null;
        arr[count - 1][1] = null;
        count--;
    }

    static void DeleteGivenPosition(String[][] arr, Scanner input){
        if (count == 0) {
            System.out.println("Array is empty");
            return;
        }

        System.out.print("input position: ");
        if(input.hasNextInt()){
            int position = input.nextInt();
            input.nextLine();

            if(position >= 1 && position <= MAX){
                int index = position - 1;
                
                if (arr[index][0] == null) {
                    System.out.println("Position is already empty");
                    return;
                }

                // for (int i = index; i < count - 1; i++) {
                //     arr[i][0] = arr[i + 1][0];
                //     arr[i][1] = arr[i + 1][1];
                // }

                arr[position - 1][0] = null;
                arr[position -1 ][1] = null;
                count--;

            } else {
                System.out.println("range out of the bound");
            }
        } else {
            System.out.println("only accept int");
            input.nextLine();
        }
    }

    static void DeleteFromEnd(String[][] arr){
        if (count == 0) {
            System.out.println("Array is empty");
            return;
        }

        arr[count - 1][0] = null;
        arr[count - 1][1] = null;
        count--;
    }

    static void DeleteFirstOccurence(String[][] arr, Scanner input){
        if (count == 0) {
            System.out.println("Array is empty");
            return;
        }

        System.out.print("input nim will be delete: ");
        String nim = input.nextLine();
        String nimReplace = nim.replaceAll("\\s","");
        
        int positionToDelete = -1;
        boolean found = false;

        for(int i = 0; i < count; i++){
            if(arr[i][1] != null){
                String replace = arr[i][1].replaceAll("\\s","");
                if(replace.equalsIgnoreCase(nimReplace)) {
                    positionToDelete = i;
                    found = true;
                    break;
                }
            }
        }

        if(found){
            for (int i = positionToDelete; i < count - 1; i++) {
                arr[i][0] = arr[i + 1][0];
                arr[i][1] = arr[i + 1][1];
            }
            arr[count - 1][0] = null;
            arr[count - 1][1] = null;
            count--;
        } else {
            System.out.println("nim not found");
        }
    }

    static void ShowData(String[][] arr)
    {   
        
        System.out.println("Length array / total data: " + count);
        for(int i = 0; i < arr.length; i++){
            if (arr[i][0] != null) {
                System.out.println((i + 1) +" : "+ arr[i][0] + " || " + arr[i][1]);
            } else {
                System.out.println((i + 1) +" : [Empty]");
            }
        }
    }
}